tests/test_resolvers.py

```diff
@@ -0,0 +1,237 @@
+        async def mock_resolve_side_effect(hostname, record_type):
+            if record_type == "A":
+                return mock_answer_a
+            elif record_type == "AAAA":
+                return mock_answer_aaaa
+            else:
+                raise dns.resolver.NXDOMAIN()
+
+        mock_resolve.side_effect = mock_resolve_side_effect
+
+        ma = Multiaddr("/dnsaddr/example.com")
+        result = await dns_resolver.resolve(ma)
+        assert len(result) == 1
+        assert result[0].protocols()[0].name == "ip4"
+        assert result[0].value_for_protocol(result[0].protocols()[0].code) == "127.0.0.1"
+
+
+@pytest.mark.trio
+async def test_resolve_dns_addr_with_mixed_quotes(dns_resolver):
+    """Test resolving DNS records with mixed quotes."""
+    # Create mock DNS answer for A record (IPv4)
+    mock_answer_a = AsyncMock()
+    mock_rdata_a = AsyncMock()
+    mock_rdata_a.address = "127.0.0.1"
+    mock_answer_a.__iter__.return_value = [mock_rdata_a]
+
+    # Create mock DNS answer for AAAA record (IPv6) - return empty to avoid conflicts
+    mock_answer_aaaa = AsyncMock()
+    mock_answer_aaaa.__iter__.return_value = []
+
+    with patch.object(dns_resolver._resolver, 'resolve') as mock_resolve:
+        # Configure the mock to return different results based on record type
+        async def mock_resolve_side_effect(hostname, record_type):
+            if record_type == "A":
+                return mock_answer_a
+            elif record_type == "AAAA":
+                return mock_answer_aaaa
+            else:
+                raise dns.resolver.NXDOMAIN()
+
+        mock_resolve.side_effect = mock_resolve_side_effect
+
+        ma = Multiaddr("/dnsaddr/example.com")
+        result = await dns_resolver.resolve(ma)
+        assert len(result) == 1
+        assert result[0].protocols()[0].name == "ip4"
+        assert result[0].value_for_protocol(result[0].protocols()[0].code) == "127.0.0.1"
+
+
+@pytest.mark.trio
+async def test_resolve_cancellation_with_error():
+    """Test that DNS resolution can be cancelled."""
+    ma = Multiaddr("/dnsaddr/nonexistent.example.com")
+    signal = trio.CancelScope()  # type: ignore[call-arg]
+    signal.cancelled_caught = True
+    dns_resolver = DNSResolver()
+
+    # Mock the DNS resolver to simulate a slow lookup that can be cancelled
+    async def slow_dns_resolve(*args, **kwargs):
+        await trio.sleep(0.5)  # Long sleep to allow cancellation
+        raise dns.resolver.NXDOMAIN("Domain not found")
+
+    with patch.object(dns_resolver._resolver, 'resolve', side_effect=slow_dns_resolve):
+        # Start resolution in background and cancel it
+        async with trio.open_nursery() as nursery:
+            # Start the resolution
+            nursery.start_soon(dns_resolver.resolve, ma, {"signal": signal})
+
+            # Cancel after a short delay
+            await trio.sleep(0.1)
+            signal.cancel()
+
+            # The nursery should handle the cancellation
```

Comment:
Should there be some assertion here? Or it will throw an exception if it is not handled correctly?

Reply:
Yes thanks, assertion was missing. 
The test now asserts that signal.cancel_called is True, confirming that cancellation was actually triggered.